File: leetcode/2001-3000/2901-3000/2981-find-longest-special-substring-that-occurs-thrice-i/FindLongestSpecialSubstringThatOccursThriceI.java
import java.util.function.Function;

public class FindLongestSpecialSubstringThatOccursThriceI {
  public static void main(String[] args) {
    // String s = "aaaa";
    String s = "aaababacaaaa";

    System.out.println((new Solution()).maximumLength(s));
  }
}

class Solution {
  public int maximumLength(String s) {
    int left = 0;
    int right = s.length();

    Function<Integer, Boolean> hasSubstringOccuringThrice = (size) -> {
      int[] repeats = new int[26];

      for (int i = 0; i < s.length();) {
        int j = i + 1;

        while (j < s.length() && s.charAt(j) == s.charAt(j - 1)) {
          j++;
        }

        int repeat = j - i - size + 1;

        int index = s.charAt(i) - 'a';

        repeats[index] += Math.max(0, repeat);

        if (repeats[index] >= 3) {
          return true;
        }

        i = j;
      }

      return false;
    };

    while (left < right) {
      int middle = (left + right + 1) / 2;

      if (hasSubstringOccuringThrice.apply(middle)) {
        left = middle;
      } else {
        right = middle - 1;
      }
    }

    return left == 0 ? -1 : left;
  }
}

File: leetcode/2001-3000/2901-3000/2981-2990/2981-find-longest-special-substring-that-occurs-thrice-i/FindLongestSpecialSubstringThatOccursThriceI.ts
function maximumLength(s: string): number {
  let left = 0;
  let right = s.length;
  let length = s.length;

  const hasSubstring = (size: number): boolean => {
    let i: number = 0;
    let j: number = 0;
    const repeats = {};

    while (i < length) {

      while (j < length && s[j] === s[j - 1]) {
        j++;
      }

      // Number of times
      const times = Math.max(0, j - i - size + 1);
      repeats[s[i]] = (repeats[s[i]] ?? 0) + times;

      if (repeats[s[i]] >= 3) {
        return true;
      }

      // Preparation for the next round
      i = j;
      j = i + 1;
    }

    return false;
  }

  while (left < right) {
    const middle: number = (left + right + 1) >> 1;

    if (hasSubstring(middle)) {
      left = middle;
    } else {
      right = middle - 1;
    }
  }

  return left === 0 ? -1 : left;
};

const s: string = "ahohhhvggh";
// const s: string = "abcdefabcdefaabbccddeeffaaabbbdaaaa";
// const s = "aaaa";

console.log(maximumLength(s));

